### PR TITLE
[BlockDAO] new chain db file implementation

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: blockchain/blockdao/filedao.go [hunk 1]
```diff
@@ -69,7 +74,7 @@ func NewFileDAO(compressBlock bool, cfg config.DB) (FileDAO, error) {
 	if err != nil {
 		return nil, err
 	}
-	return createFileDAO(legacyFd)
+	return createFileDAO(legacyFd, nil)
```
Comment (dustinxie) [inline blockchain/blockdao/filedao.go:77]: still use legacy file for now
next PR will switch knob to use combination of legacy and new file

File: blockchain/blockdao/filedao_new.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+	stagingDataNS = "stg"
+)
+
+var (
+	bottomHeightKey = []byte("bh")
+)
+
+type (
+	// FileDAONew has extra methods for chain db file after file split activation at Ithaca height
+	FileDAONew interface {
```
Comment (yutongp) [inline blockchain/blockdao/filedao_new.go:45]: i would say this is not a good name. how would you name the next version of filedao, filedaonewnew?

Comment (dustinxie) [reply blockchain/blockdao/filedao_new.go:45]: ok, how about fileDAOV2?

File: blockchain/blockdao/filedao_v2_test.go [hunk 3]
```diff
@@ -323,15 +340,21 @@ func TestNewFdStart(t *testing.T) {
...
 	blk, _ := builder.
 		SetHeight(height).
 		SetPrevBlockHash(h).
 		SetReceipts([]*action.Receipt{
-			&action.Receipt{
-				Status:      1,
-				BlockHeight: height,
-				ActionHash:  h,
-			},
+			r.AddTransactionLogs(&action.TransactionLog{
```
Comment (dustinxie) [inline blockchain/blockdao/filedao_v2_test.go:352]: add a transaction log into test block

File: blockchain/blockdao/filedao_v2.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	stagingDataNS = "stg"
+)
+
+var (
+	bottomHeightKey = []byte("bh")
+)
+
+type (
+	// FileDAOv2 has extra methods for chain db file after file split activation at Ithaca height
+	FileDAOv2 interface {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2.go:45]: keen catch
turns out Bottom() ContainsHeight() is only used inside v2, so no need to expose this interface

File: blockchain/blockdao/filedao_v2.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+		bottomHeight:  bottom,
+		cfg:           cfg,
+		blkCache:      cache.NewThreadSafeLruCache(16),
+		kvStore:       kvStore,
+		batch:         batch.NewBatch(),
+	}
+
+	// write the bottom height
+	ctx := context.Background()
+	if err := fd.kvStore.Start(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2.go:82]: keen catch
moved to Start()

File: blockchain/blockdao/filedao_v2.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2.go:45]: i think no need, but can do if you think that's necessary, in next PR?

File: blockchain/blockdao/filedao_v2.go [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+		compressBlock: cfg.CompressData,
+		bottomHeight:  bottom,
+		cfg:           cfg,
+		blkCache:      cache.NewThreadSafeLruCache(16),
+		kvStore:       kvStore,
+		batch:         batch.NewBatch(),
+	}
+
+	// write the bottom height
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2.go]: fixed in latest push

File: pkg/compress/compress.go [hunk 7]
```diff
@@ -38,3 +40,17 @@ func Decompress(data []byte) ([]byte, error) {
 	r.Close()
 	return ioutil.ReadAll(r)
 }
+
+// CompV2 uses Snappy to compress the input bytes
+func CompV2(data []byte) ([]byte, error) {
```
Comment (yutongp) [inline pkg/compress/compress.go:45]: maybe just CompressSnappy

File: blockchain/blockdao/filedao_v2_util.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	if !comp {
+		return append(value, _normal), nil
+	}
+
+	var err error
+	value, err = compress.CompV2(value)
+	if err != nil {
+		return nil, err
+	}
+	return append(value, _compressed), nil
```
Comment (yutongp) [inline blockchain/blockdao/filedao_v2_util.go:141]: shall we just put if data is compressed with what compression schema in file header?

File: blockchain/blockdao/filedao_v2_util.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+	if height <= start {
+		return 0
+	}
+	return (height - start) / blockStoreBatchSize
+}
+
+// stagingKey is the position of block in the staging buffer
+func stagingKey(height, start uint64) uint64 {
+	return (height - start) % blockStoreBatchSize
+}
```
Comment (yutongp) [inline blockchain/blockdao/filedao_v2_util.go:182]: i am not sure if this is better or let stagingbuff maintain next index is better. I actually prefer later one, but i cannot tell exactly why.

File: blockchain/blockdao/filedao_v2.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+	blockDataNS   = "bdn"
+	stagingDataNS = "stg"
+)
+
+var (
+	bottomHeightKey = []byte("bh")
+)
+
+type (
+	fileDAOv2 struct {
```
Comment (yutongp) [inline blockchain/blockdao/filedao_v2.go:44]: i think filedao maybe deserve it own package? blockdao/filedao?

**Review by yutongp — APPROVED** [review]

File: blockchain/blockdao/filedao_v2_util.go (hunk 8 above)
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2_util.go]: fixed

File: blockchain/blockdao/filedao_v2.go (hunk 10 above)
Comment (dustinxie) [reply blockchain/blockdao/filedao_v2.go]: fixed

**Pull request merged**